### PR TITLE
feat(web,server): rotate photo

### PR DESCRIPTION
**Pull request by martabal**
## Changes made in this PR

This PR adds the ability to rotate a photo when viewing it. The photo can be rotated with a new menu option or with the `⇧ + r` shortcut

To do:
- [ ] Add unit tests
- [ ] Save / use the saved orientation

## Screenshots

https://github.com/immich-app/immich/assets/74269598/7d1be92b-67c0-4aeb-be25-36c96a74a9a9

### CODE REVIEW

**Comment by bdr99** [comment]
I think the rotation direction as currently implemented in this PR is actually reversed. Typically "rotate left" means counter-clockwise and "rotate right" would be clockwise.

**Comment by martabal** [comment]
> I think the rotation direction as currently implemented in this PR is actually reversed. Typically "rotate left" means counter-clockwise and "rotate right" would be clockwise.

Indeed, thanks

**Review by jrasm91 — COMMENTED** [review]
Can you add a unit test for the server please?

File: web/src/lib/components/asset-viewer/asset-viewer.svelte [hunk 1]
```diff
@@ -256,6 +255,11 @@
     isShowActivity = !isShowActivity;
   };
 
+  const handleRotate = () => {
+    const rotateImage = new CustomEvent('rotateImage');
```
[the remark this answers is not in the document]
Comment (jrasm91) [reply web/src/lib/components/asset-viewer/asset-viewer.svelte:258]: Why does zoom image need a custom event...?

[the remark this answers is not in the document]
Comment (martabal) [reply web/src/lib/components/asset-viewer/asset-viewer.svelte:258]: So other components can change the rotation of the image.

Would this be better ?
```ts
  export const rotate = async () => {
    setZoomImageWheelState({ currentRotation: $zoomImageWheelState.currentRotation + 90 });
    try {
      await api.assetApi.updateAsset({
        id: asset.id,
        updateAssetDto: { orientation: getRotationString($zoomImageWheelState.currentRotation) },
      });
    } catch (error) {
      handleError(error, 'Unable to change orientation');
    }
  };
```

[the remark this answers is not in the document]
Comment (jrasm91) [reply web/src/lib/components/asset-viewer/asset-viewer.svelte:258]: Can't you just use the event dispatcher like everything else?

[the remark this answers is not in the document]
Comment (martabal) [reply web/src/lib/components/asset-viewer/asset-viewer.svelte:258]: How ? Can the dispatcher send an event from a parent to a child component ?

[the remark this answers is not in the document]
Comment (jrasm91) [reply web/src/lib/components/asset-viewer/asset-viewer.svelte:258]: You can pass data from a parent to a child just by using props.

[the remark this answers is not in the document]
Comment (martabal) [reply web/src/lib/components/asset-viewer/asset-viewer.svelte:258]: What's the difference with [this](https://github.com/immich-app/immich/pull/5151#discussion_r1412872995) ?
```svelte
let rotatePhotoviewer: () => Promise<void>;
<PhotoViewer bind:rotate={rotatePhotoviewer} {asset} on:close={closeViewer} />
```

File: web/src/lib/components/asset-viewer/photo-viewer.svelte [hunk 2]
```diff
@@ -117,23 +195,32 @@
...
 >
   {#await loadAssetData({ loadOriginal: false })}
     <LoadingSpinner />
   {:then}
-    <div bind:this={imgElement} class="h-full w-full">
+    <div
+      bind:this={imgElement}
+      class="duration-500"
+      style={asset.exifInfo?.orientation
+        ? `transform: rotate(${orientationToRotation(asset.exifInfo?.orientation)}deg);`
```
Comment (jrasm91) [inline web/src/lib/components/asset-viewer/photo-viewer.svelte:214]: Pretty sure this will conflict with thumbnails which already take rotation into account when they are generated.

Comment (bo0tzz) [reply web/src/lib/components/asset-viewer/photo-viewer.svelte:214]: Should we trigger thumbnail generation when rotation changes, and then just always use the thumbnail as is?

Comment (martabal) [reply web/src/lib/components/asset-viewer/photo-viewer.svelte:214]: Yeah, currently all the webUI (with the correct width / height / animation) is done. Now, I'm trying to figuring out how to use correctly the orientation information. 

Another issue is live photos (google disables the live part when the image orientation is changed).

File: server/src/domain/asset/asset.service.ts [hunk 3]
```diff
@@ -333,6 +333,9 @@ export class AssetService {
     await this.updateMetadata({ id, description, dateTimeOriginal, latitude, longitude, orientation });
 
     const asset = await this.assetRepository.save({ id, ...rest });
+    if (orientation) {
```
Comment (jrasm91) [inline server/src/domain/asset/asset.service.ts:336]: Job scheduling can be complicated. Personally, I would see if there is a way to implement this in the job service itself. Basically, in the `onDone`. What should happen is what if the sidecar write job writes the orientation tag _then_ it should automatically trigger rotation.

File: server/src/infra/repositories/media.repository.ts [hunk 4]
```diff
@@ -27,10 +35,18 @@ export class MediaRepository implements IMediaRepository {
...
+      options.format === 'webp'
+        ? orientation
+          ? orientationExifToDegrees(orientation.toString() as Orientation)
+          : undefined
+        : undefined;
     await sharp(input, { failOn: 'none' })
       .pipelineColorspace(options.colorspace === Colorspace.SRGB ? 'srgb' : 'rgb16')
       .resize(options.size, options.size, { fit: 'outside', withoutEnlargement: true })
-      .rotate()
+      .rotate(rotation)
```
Comment (jrasm91) [inline server/src/infra/repositories/media.repository.ts:48]: Is this change required because it needs to use the overridden value from the sidecar file now? (as opposed to the value in the original file itself?

Comment (martabal) [reply server/src/infra/repositories/media.repository.ts:48]: Yes, exactly
